nilearn/decomposition/base.py

```diff
@@ -0,0 +1,533 @@
+         combination in whole-brain voxel space
+
+        Parameters
+        ----------
+        loadings: list of numpy array (n_samples x n_components)
+            Component signals to tranform back into voxel signals
+
+        Returns
+        ----------
+        reconstructed_imgs: list of nibabel.Nifti1Image
+           For each loading, reconstructed Nifti1Image
+        """
+        if not hasattr(self, 'components_'):
+            ValueError('Object has no components_ attribute. This is either '
+                       'because fit has not been called or because'
+                       '_DecompositionEstimator has direcly been used')
+        self._check_components_()
+        components_img_ = self.masker_.inverse_transform(self.components_)
+        nifti_maps_masker = NiftiMapsMasker(
+            components_img_, self.masker_.mask_img_,
+            resampling_target='maps')
+        nifti_maps_masker.fit()
+        # XXX: dealing properly with 2D/ list of 2D data?
+        return [nifti_maps_masker.inverse_transform(loading)
+                for loading in loadings]
+
+    def _raw_score(self, data, per_component=True):
+        return explained_variance(data, self.components_,
+                                  per_component=per_component)
+
+    def score(self, imgs, confounds=None, per_component=True):
+        """Score function based on explained variance on imgs.
+
+        Should only be used by DecompositionEstimator derived classes
+
+        Parameters
+        ----------
+        imgs: iterable of Niimg-like objects
+            See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+            Data to be scored
+
+        confounds: CSV file path or 2D matrix
+            This parameter is passed to nilearn.signal.clean. Please see the
+            related documentation for details
+
+        per_component: boolean,
+            Specify whether the explained variance ratio is desired for each
+             map or for the global set of components
+
+        Returns
+        -------
+        score: ndarray or float,
+            Holds the score for each subjects. Score is two dimensional
+             if per_component is True. First dimension
+            is squeezed if the number of subjects is one
+        """
+        if not isinstance(imgs, tuple) and not isinstance(imgs, list):
+            imgs = [imgs]
+        if confounds is None:
+            confounds = [None] * len(imgs)
+        if per_component:
+            score = np.zeros((len(imgs), self.n_components))
+        else:
+            score = np.zeros(len(imgs))
+        for i, (img, confound) in enumerate(zip(imgs, confounds)):
+            data = self.masker_.transform(img, confound)
+            score[i] = self._raw_score(data, per_component=per_component)
+        return score if len(imgs) > 1 else score[0]
+
+def explained_variance(X, components, per_component=True):
+    """Score function based on explained variance
```

Comment:
kind of tautological with the score() function of the base class #descriminability